### PR TITLE
[WIP] Migrate to PDO using compatibility layer

### PR DESCRIPTION
**Pull request by Sobak**
I tried to start a migration several times in the past and gave up each time - this codebase is too critical and  messy to do this reliably. This time I changed my approach - I decided to write thin compatibility layer between MDB2 and PDO APIs. Most of the existing code should remain untouched with only exceptions being:

- calls to `numRows()` which PDO has no counterpart of
- errors handling: application-wide exception handler will need to be added

This is obviously still work in progress but I'd love to hear any opinions on general idea of migrating to PDO, suggested approach and its implementation so far.

### CODE REVIEW

File: include/prepend.php [hunk 1]
```diff
@@ -45,23 +45,20 @@
...
 $basedir = $site_data['basedir'];
 define('BUG_PATCHTRACKER_TMPDIR', $site_data['patch_tmp']);
-define('DATABASE_DSN', "{$site_data['db_extension']}://{$site_data['db_user']}:{$site_data['db_pass']}@{$site_data['db_host']}/{$site_data['db']}");
+define('DATABASE_DSN', "mysql:host={$site_data['db_host']};dbname={$site_data['db']};charset=utf8");
 
 /**
  * Obtain the functions and variables used throughout the bug system
  */
 require_once "{$ROOT_DIR}/include/functions.php";
+require 'classes/bug_pdo.php';
```
Comment (johannes) [inline include/prepend.php]: Is there any reason not to use require_once and the `$ROOT_DIR` as in the line above?

**Comment by heiglandreas** [comment]
As no one is actively complaining I'd say it's a good idea to move to PDO.

What was your plan on resolving the ```numRows```-issues? 

And why would there need to be an application-wide exception-handler? Without having had a deeper look into the code: New exceptions can only occur within the wrapper that should then handle those exceptions so they are not propagating outside of the handler… Or am I missing a point?

File: include/prepend.php (hunk 1 above)
Comment (Sobak) [inline include/prepend.php]: No, in fact there isn't any. I'll change that to be consistent

**Comment by Sobak** [comment]
> What was your plan on resolving the `numRows`-issues?

These are already resolved. AFAIK there's no occurecence of `numRows` left. I replaced them with apropriate checks, depending on the actual case.

> And why would there need to be an application-wide exception-handler? Without having had a deeper look into the code: New exceptions can only occur within the wrapper that should then handle those exceptions so they are not propagating outside of the handler… Or am I missing a point?

In fact it was an attempt to directly map PEAR error handler, but you might be right. I'll need to check that.

**Comment by heiglandreas** [comment]
Sorry, I didn't check in the patch but it actually looks like all occurrences of ```numRows``` are replaced in the patch! Thanks! So consider that as obsolete :wink:

**Comment by philip** [comment]
+1 for changing to PDO :)

**Comment by rlerdorf** [comment]
A version of this has been merged. Thanks

**Pull request closed without merging**